### PR TITLE
[27] Text preprocessing and cluster keyword generation

### LINKED ISSUE #27: Generate keywords for each cluster
Add functionality to extract keywords and keyphrases characterising app clusters. It should work as follows.

Inputs:
- App text data (eg descriptions)
- App cluster integer labels

Processing:
- Clean text data
- Tokenise text data (eg words, or phrases)
- Use tf-idf scores to find the most salient tokens

Outputs:
- Top-n words or phrases for each cluster

### PR DESCRIPTION
**Pull request by beingkk**
Closes #27 

New files/modules:
- `mapping_parenting_tech.analysis.cluster_analysis_utils`: Added functionality to generate cluster keywords using TF-IDF vectors
- `mapping_parenting_tech.utils.text_preprocessing_utils`: Added a module with functions that carry out the required text preprocessing steps. This will also be useful when doing topic modelling of review text data.
- `prototyping/cluster_keywords.py`: Example how to use these modules

NB: Haven't yet added unit tests for these functions (will create a new issue to do that later)

---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

### CODE REVIEW

**Comment by beingkk** [comment]
Hi @mathickman , hopefully the keyword generator will be useful. Looks like you already did a great job at characterising the app clusters though! Nonetheless, the text preprocessing module should be useful in further review analysis as well.

Happy to hear suggestions to improve the code (perhaps some parts can be further modularised?)

FYI comparison between your labels and the keywords:
```
{'parent support / keeping in touch / photo sharing': ['baby',
  'pregnancy',
  'contraction',
  'birth',
  'tracker',
  'pamper',
  'milestone',
  'mom',
  'date',
  'name'],
 "baby 'habit' tracking, inc. breastfeeding": ['baby',
  'diaper',
  'tracker',
  'parent',
  'feeding',
  'sleep',
  'breast',
  'milk',
  'growth',
  'milestone'],
 'baby sleep': ['sleep',
  'baby',
  'lullaby',
  'flashlight',
  'sound',
  'noise',
  'light',
  'night',
  'music',
  'palette'],
 'young games': ['kid',
  'animal',
  'puzzle',
  'toddler',
  'dinosaur',
  'car',
  'baby',
  'tutotoon',
  'pet',
  'truck'],
 'period and fertility tracking': ['ovulation',
  'fertility',
  'cycle',
  'tracker',
  'pregnancy',
  'calendar',
  'symptom',
  'woman',
  'exercise',
  'health'],
 'ABCs': ['letter',
  'kid',
  'alphabet',
  'word',
  'phonic',
  'toddler',
  'puzzle',
  'animal',
  'sound',
  'tracing'],
 'stories': ['story',
  'kid',
  'tale',
  'fairy',
  'pig',
  'peppa',
  'reading',
  'audiobook',
  'shark',
  'character'],
 'maths games + science education': ['math',
  'kid',
  'subtraction',
  'multiplication',
  'toddler',
  'counting',
  'dice',
  'kindergarten',
  'table',
  'calculator'],
 'learning to read and write': ['word',
  'spelling',
  'letter',
  'kid',
  'phonic',
  'sight',
  'reading',
  'vowel',
  'dolch',
  'alphabet'],
 'baby/toddler educational games': ['kid',
  'toddler',
  'puzzle',
  'baby',
  'shape',
  'animal',
  'object',
  'letter',
  'kindergarten',
  'parent'],
 'drawing / colouring in': ['coloring',
  'kid',
  'drawing',
  'toddler',
  'painting',
  'glitter',
  'bead',
  'paint',
  'animal',
  'brush'],
 'music apps, inc for toddlers': ['song',
  'music',
  'piano',
  'kid',
  'guitar',
  'rhyme',
  'chord',
  'instrument',
  'nursery',
  'baby'],
 'learning [foreign?] language': ['word',
  'phrase',
  'speaker',
  'kid',
  'lesson',
  'vocabulary',
  'pronunciation',
  'grammar',
  'funeasylearn',
  'learner'],
 'Russian': ['sd',
  'ubbu',
  'supportcatdonutcom',
  'android',
  'bu',
  'votre',
  'ou',
  'pour',
  'mindfulness',
  'duzakangmailcom'],
 'medical': ['drug',
  'disease',
  'medicine',
  'nursing',
  'treatment',
  'emergency',
  'diagnosis',
  'word',
  'patient',
  'nurse'],
 'student learning': ['school',
  'student',
  'blockchain',
  'flashcard',
  'parent',
  'teacher',
  'homework',
  'timetable',
  'programming',
  'attendance'],
 'driving test': ['test',
  'theory',
  'hazard',
  'dvsa',
  'driver',
  'highway',
  'revision',
  'lgv',
  'driving',
  'perception'],
 'medical education': ['anatomy',
  'nursing',
  'acupuncture',
  'surgery',
  'patient',
  'exam',
  'histology',
  'meridian',
  'doctor',
  'usmle'],
 "medical 'organising'": ['medication',
  'skin',
  'pill',
  'patient',
  'prescription',
  'health',
  'reminder',
  'medicine',
  'skincare',
  'appointment'],
 'positivity/self development': ['quote',
  'self',
  'happiness',
  'affirmation',
  'mindset',
  'motivation',
  'success',
  'thinking',
  'cv',
  'habit'],
 'medical/health support (for adults)': ['recipe',
  'heart',
  'blood',
  'health',
  'pressure',
  'food',
  'pollen',
  'weight',
  'ecg',
  'rate'],
 'meditation/mindfulness/rest': ['meditation',
  'sleep',
  'sound',
  'chakra',
  'mindfulness',
  'angel',
  'deck',
  'music',
  'oracle',
  'hypnosis'],
 'outdoor / mapping': ['map',
  'route',
  'gps',
  'waypoint',
  'gpx',
  'topo',
  'hiking',
  'elevation',
  'trail',
  'property'],
 'web safety': ['kid',
  'location',
  'parent',
  'chore',
  'gps',
  'locator',
  'tracker',
  'alert',
  'usage',
  'notification'],
 'being organised/time management': ['calendar',
  'task',
  'event',
  'habit',
  'planner',
  'widget',
  'reminder',
  'agenda',
  'shift',
  'schedule'],
 'engineering/physics': ['flight',
  'calculator',
  'airport',
  'pipe',
  'calculation',
  'sky',
  'angle',
  'constellation',
  'airline',
  'star'],
 'TBC': ['anxiety',
  'psychology',
  'health',
  'fact',
  'therapy',
  'mood',
  'depression',
  'emotion',
  'stress',
  'brain'],
 'religion': ['bible',
  'verse',
  'prayer',
  'testament',
  'church',
  'god',
  'commentary',
  'sermon',
  'scripture',
  'word']}
```

File: mapping_parenting_tech/analysis/prototyping/cluster_keywords.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+# Get this file from S3
+input_file = PROJECT_DIR / "inputs/data/misc/HLE app reviewing - apps_to_review.csv"
+
+# %% [markdown]
+# ## Prerequisites
+
+# %%
+# Create a spacy model instance
+nlp = spacy.load("en_core_web_sm")
```
Comment (mathickman) [inline mapping_parenting_tech/analysis/prototyping/cluster_keywords.py:36]: This threw an error for me:
```OSError                                   Traceback (most recent call last)
Input In [2], in <module>
      1 # Create a spacy model instance
----> 2 nlp = spacy.load("en_core_web_sm")
      4 # Get the input data
      5 apps_df = pd.read_csv(input_file)

File ~/miniconda3/envs/mapping_parenting_tech/lib/python3.8/site-packages/spacy/__init__.py:51, in load(name, vocab, disable, exclude, config)
     30 def load(
     31     name: Union[str, Path],
     32     *,
   (...)
     36     config: Union[Dict[str, Any], Config] = util.SimpleFrozenDict(),
     37 ) -> Language:
     38     """Load a spaCy model from an installed package or a local path.
     39 
     40     name (str): Package name or model path.
   (...)
     49     RETURNS (Language): The loaded nlp object.
     50     """
---> 51     return util.load_model(
     52         name, vocab=vocab, disable=disable, exclude=exclude, config=config
     53     )

File ~/miniconda3/envs/mapping_parenting_tech/lib/python3.8/site-packages/spacy/util.py:427, in load_model(name, vocab, disable, exclude, config)
    425 if name in OLD_MODEL_SHORTCUTS:
    426     raise IOError(Errors.E941.format(name=name, full=OLD_MODEL_SHORTCUTS[name]))  # type: ignore[index]
--> 427 raise IOError(Errors.E050.format(name=name))

OSError: [E050] Can't find model 'en_core_web_sm'. It doesn't seem to be a Python package or a valid path to a data directory.```

Comment (beingkk) [reply mapping_parenting_tech/analysis/prototyping/cluster_keywords.py:36]: Oh, you might need to download and install that particular spacy model. You can try in the terminal:
```
pip install https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.2.0/en_core_web_sm-3.2.0.tar.gz
```

I will also add this requirement to the `requirements.txt` file.

If the problem persists, perhaps can re-install spacy, or check [this issue](https://github.com/explosion/spaCy/issues/4577)...

Comment (mathickman) [reply mapping_parenting_tech/analysis/prototyping/cluster_keywords.py:36]: Done and working fine

**Review by mathickman — APPROVED** [review]

**Pull request merged**